// @vercel/microfrontends/dist/microfrontends/utils.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/config/microfrontends/utils/index.ts
var utils_exports = {};
__export(utils_exports, {
  findConfig: () => findConfig,
  generateDefaultAssetPrefixFromName: () => generateDefaultAssetPrefixFromName,
  getPossibleConfigurationFilenames: () => getPossibleConfigurationFilenames,
  inferMicrofrontendsLocation: () => inferMicrofrontendsLocation
});
module.exports = __toCommonJS(utils_exports);

// src/config/microfrontends/utils/find-config.ts
var import_node_fs = __toESM(require("fs"), 1);
var import_node_path = require("path");

// src/config/microfrontends/utils/get-config-file-name.ts
var DEFAULT_CONFIGURATION_FILENAMES = [
  "microfrontends.json",
  "microfrontends.jsonc"
];
function getPossibleConfigurationFilenames({
  customConfigFilename
}) {
  if (customConfigFilename) {
    if (!customConfigFilename.endsWith(".json") && !customConfigFilename.endsWith(".jsonc")) {
      throw new Error(
        `Found VC_MICROFRONTENDS_CONFIG_FILE_NAME but the name is invalid. Received: ${customConfigFilename}. The file name must end with '.json' or '.jsonc'. It's also possible for the env var to include the path, eg microfrontends-dev.json or /path/to/microfrontends-dev.json.`
      );
    }
    return Array.from(
      /* @__PURE__ */ new Set([customConfigFilename, ...DEFAULT_CONFIGURATION_FILENAMES])
    );
  }
  return DEFAULT_CONFIGURATION_FILENAMES;
}

// src/config/microfrontends/utils/find-config.ts
function findConfig({
  dir,
  customConfigFilename
}) {
  for (const filename of getPossibleConfigurationFilenames({
    customConfigFilename
  })) {
    const maybeConfig = (0, import_node_path.join)(dir, filename);
    if (import_node_fs.default.existsSync(maybeConfig)) {
      return maybeConfig;
    }
  }
  return null;
}

// src/config/microfrontends/utils/hash-application-name.ts
var import_md5 = __toESM(require("md5"), 1);
function hashApplicationName(name) {
  if (!name) {
    throw new Error("Application name is required to generate hash");
  }
  return (0, import_md5.default)(name).substring(0, 6).padStart(6, "0");
}

// src/config/microfrontends/utils/generate-default-asset-prefix.ts
var PREFIX = "vc-ap";
function generateDefaultAssetPrefixFromName({
  name
}) {
  if (!name) {
    throw new Error("Name is required to generate an asset prefix");
  }
  return `${PREFIX}-${hashApplicationName(name)}`;
}

// src/config/microfrontends/utils/infer-microfrontends-location.ts
var import_node_fs2 = require("fs");
var import_node_path2 = require("path");
var import_fast_glob = __toESM(require("fast-glob"), 1);
var import_jsonc_parser = require("jsonc-parser");

// src/bin/logger.ts
function debug(...args) {
  if (process.env.MFE_DEBUG) {
    console.log(...args);
  }
}
function info(...args) {
  console.log(...args);
}
function warn(...args) {
  console.warn(...args);
}
function error(...args) {
  console.error(...args);
}
var logger = {
  debug,
  info,
  warn,
  error
};

// src/config/errors.ts
var MicrofrontendError = class extends Error {
  constructor(message, opts) {
    super(message, { cause: opts?.cause });
    this.name = "MicrofrontendsError";
    this.source = opts?.source ?? "@vercel/microfrontends";
    this.type = opts?.type ?? "unknown";
    this.subtype = opts?.subtype;
    Error.captureStackTrace(this, MicrofrontendError);
  }
  isKnown() {
    return this.type !== "unknown";
  }
  isUnknown() {
    return !this.isKnown();
  }
  /**
   * Converts an error to a MicrofrontendsError.
   * @param original - The original error to convert.
   * @returns The converted MicrofrontendsError.
   */
  static convert(original, opts) {
    if (opts?.fileName) {
      const err = MicrofrontendError.convertFSError(original, opts.fileName);
      if (err) {
        return err;
      }
    }
    if (original.message.includes(
      "Code generation from strings disallowed for this context"
    )) {
      return new MicrofrontendError(original.message, {
        type: "config",
        subtype: "unsupported_validation_env",
        source: "ajv"
      });
    }
    return new MicrofrontendError(original.message);
  }
  static convertFSError(original, fileName) {
    if (original instanceof Error && "code" in original) {
      if (original.code === "ENOENT") {
        return new MicrofrontendError(`Could not find "${fileName}"`, {
          type: "config",
          subtype: "unable_to_read_file",
          source: "fs"
        });
      }
      if (original.code === "EACCES") {
        return new MicrofrontendError(
          `Permission denied while accessing "${fileName}"`,
          {
            type: "config",
            subtype: "invalid_permissions",
            source: "fs"
          }
        );
      }
    }
    if (original instanceof SyntaxError) {
      return new MicrofrontendError(
        `Failed to parse "${fileName}": Invalid JSON format.`,
        {
          type: "config",
          subtype: "invalid_syntax",
          source: "fs"
        }
      );
    }
    return null;
  }
  /**
   * Handles an unknown error and returns a MicrofrontendsError instance.
   * @param err - The error to handle.
   * @returns A MicrofrontendsError instance.
   */
  static handle(err, opts) {
    if (err instanceof MicrofrontendError) {
      return err;
    }
    if (err instanceof Error) {
      return MicrofrontendError.convert(err, opts);
    }
    if (typeof err === "object" && err !== null) {
      if ("message" in err && typeof err.message === "string") {
        return MicrofrontendError.convert(new Error(err.message), opts);
      }
    }
    return new MicrofrontendError("An unknown error occurred");
  }
};

// src/config/microfrontends/utils/infer-microfrontends-location.ts
var configCache = {};
function findPackageWithMicrofrontendsConfig({
  repositoryRoot,
  applicationContext,
  customConfigFilename
}) {
  const applicationName = applicationContext.name;
  logger.debug(
    "[MFE Config] Searching repository for configs containing application:",
    applicationName
  );
  try {
    const microfrontendsJsonPaths = import_fast_glob.default.globSync(
      `**/{${getPossibleConfigurationFilenames({ customConfigFilename }).join(",")}}`,
      {
        cwd: repositoryRoot,
        absolute: true,
        onlyFiles: true,
        followSymbolicLinks: false,
        ignore: ["**/node_modules/**", "**/.git/**"]
      }
    );
    logger.debug(
      "[MFE Config] Found",
      microfrontendsJsonPaths.length,
      "config file(s) in repository"
    );
    const matchingPaths = [];
    for (const microfrontendsJsonPath of microfrontendsJsonPaths) {
      if (doesApplicationExistInConfig(microfrontendsJsonPath, applicationName)) {
        matchingPaths.push(microfrontendsJsonPath);
      }
    }
    logger.debug(
      "[MFE Config] Total matching config files:",
      matchingPaths.length
    );
    if (matchingPaths.length > 1) {
      throw new MicrofrontendError(
        `Found multiple \`microfrontends.json\` files in the repository referencing the application "${applicationName}", but only one is allowed.
${matchingPaths.join("\n  \u2022 ")}`,
        { type: "config", subtype: "inference_failed" }
      );
    }
    if (matchingPaths.length === 0) {
      if (repositoryRoot && doesMisplacedConfigExist(
        repositoryRoot,
        applicationName,
        customConfigFilename
      )) {
        logger.debug(
          "[MFE Config] Found misplaced config in wrong .vercel directory in repository"
        );
        const misplacedConfigPath = (0, import_node_path2.join)(
          repositoryRoot,
          ".vercel",
          customConfigFilename || "microfrontends.json"
        );
        throw new MicrofrontendError(
          `Unable to automatically infer the location of the \`microfrontends.json\` file.

A microfrontends config was found in the \`.vercel\` directory at the repository root: ${misplacedConfigPath}
However, in a monorepo, the config file should be placed in the \`.vercel\` directory in your application directory instead.

To fix this:
1. If using \`vercel link\`, run it with \`vercel link --repo\` to handle monorepos, or run \`vercel microfrontends pull --cwd=<application-directory>\` to make sure it pulls the \`microfrontends.json\` file to the correct location
2. If manually defined, move the config file to the \`.vercel\` directory in your application
3. Alternatively, set the VC_MICROFRONTENDS_CONFIG environment variable to the correct path

For more information, see: https://vercel.com/docs/cli/project-linking`,
          { type: "config", subtype: "inference_failed" }
        );
      }
      let additionalErrorMessage = "";
      if (microfrontendsJsonPaths.length > 0) {
        if (!applicationContext.projectName) {
          additionalErrorMessage = `

If the name in package.json (${applicationContext.packageJsonName}) differs from your Vercel Project name, set the \`packageName\` field for the application in \`microfrontends.json\` to ensure that the configuration can be found locally.`;
        } else {
          additionalErrorMessage = `

Names of applications in \`microfrontends.json\` must match the Vercel Project name (${applicationContext.projectName}).`;
        }
      }
      throw new MicrofrontendError(
        `Could not find a \`microfrontends.json\` file in the repository that contains the "${applicationName}" application.${additionalErrorMessage}

If your Vercel Microfrontends configuration is not in this repository, you can use the Vercel CLI to pull the Vercel Microfrontends configuration using the "vercel microfrontends pull" command, or you can specify the path manually using the VC_MICROFRONTENDS_CONFIG environment variable.

If your Vercel Microfrontends configuration has a custom name, ensure the VC_MICROFRONTENDS_CONFIG_FILE_NAME environment variable is set, you can pull the vercel project environment variables using the "vercel env pull" command.

If you suspect this is thrown in error, please reach out to the Vercel team.`,
        { type: "config", subtype: "inference_failed" }
      );
    }
    const [packageJsonPath] = matchingPaths;
    return (0, import_node_path2.dirname)(packageJsonPath);
  } catch (error2) {
    if (error2 instanceof MicrofrontendError) {
      throw error2;
    }
    return null;
  }
}
function inferMicrofrontendsLocation(opts) {
  const cacheKey = `${opts.repositoryRoot}-${opts.applicationContext.name}${opts.customConfigFilename ? `-${opts.customConfigFilename}` : ""}`;
  if (configCache[cacheKey]) {
    return configCache[cacheKey];
  }
  const result = findPackageWithMicrofrontendsConfig(opts);
  if (!result) {
    throw new MicrofrontendError(
      `Could not infer the location of the \`microfrontends.json\` file for application "${opts.applicationContext.name}" starting in directory "${opts.repositoryRoot}".`,
      { type: "config", subtype: "inference_failed" }
    );
  }
  configCache[cacheKey] = result;
  return result;
}
function existsSync(path) {
  try {
    (0, import_node_fs2.statSync)(path);
    return true;
  } catch (_) {
    return false;
  }
}
function doesMisplacedConfigExist(repositoryRoot, applicationName, customConfigFilename) {
  logger.debug(
    "[MFE Config] Looking for misplaced config in wrong .vercel directory"
  );
  const misplacedConfigPath = (0, import_node_path2.join)(
    repositoryRoot,
    ".vercel",
    customConfigFilename || "microfrontends.json"
  );
  return existsSync(misplacedConfigPath) && doesApplicationExistInConfig(misplacedConfigPath, applicationName);
}
function doesApplicationExistInConfig(microfrontendsJsonPath, applicationName) {
  try {
    const microfrontendsJsonContent = (0, import_node_fs2.readFileSync)(
      microfrontendsJsonPath,
      "utf-8"
    );
    const microfrontendsJson = (0, import_jsonc_parser.parse)(microfrontendsJsonContent);
    if (microfrontendsJson.applications[applicationName]) {
      logger.debug(
        "[MFE Config] Found application in config:",
        microfrontendsJsonPath
      );
      return true;
    }
    for (const [_, app] of Object.entries(microfrontendsJson.applications)) {
      if (app.packageName === applicationName) {
        logger.debug(
          "[MFE Config] Found application via packageName in config:",
          microfrontendsJsonPath
        );
        return true;
      }
    }
  } catch (error2) {
    logger.debug("[MFE Config] Error checking application in config:", error2);
  }
  return false;
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  findConfig,
  generateDefaultAssetPrefixFromName,
  getPossibleConfigurationFilenames,
  inferMicrofrontendsLocation
});
